kale/nbparser/parser.py

```diff
@@ -124,62 +116,88 @@ def parse_notebook(notebook):
+        # a block names and (possibly) some dependencies
+
+        # check existence of prev steps. They must already exist in the graph
+        for prev in tags['prev_steps']:
+            if prev not in nb_graph.nodes:
+                raise ValueError("Block %s does not exist. "
+                                 "It was defined as previous block of %s"
+                                 % (prev, tags['step_names']))
+
+        # if the cell was not tagged with a step name,
+        # add the code to the previous cell
+        if not step_name:
+            if prev_step_name == 'imports':
+                imports_block.append(c.source)
+            if prev_step_name == 'functions':
+                functions_block.append(c.source)
+            if prev_step_name == 'pipeline-parameters':
+                pipeline_parameters.append(c.source)
+            # current_block might be None in case the first cells of the
+            # notebooks have not been tagged.
+            if prev_step_name:
+                # this notebook cell will be merged to a previous one that
+                # specified a step name
+                merge_code(nb_graph, prev_step_name, c.source)
         else:
-            current_block_global = False
-            current_block_pipeline_parameters = False
-            # TODO: Taking by default first block name tag. Need specific behavior
-            current_block = tags['block_names'][0]
-            for block_name in tags['block_names']:
-                # add node to DAG, adding tags and source code of notebook cell
-                if block_name not in nb_graph.nodes:
-                    _tags = copy.deepcopy(tags)
-                    _tags.block_name = block_name
-                    nb_graph.add_node(block_name, tags=_tags, source=c.source,
-                                      ins=set(), outs=set())
-                    if tags['previous_blocks']:
-                        for block in tags['previous_blocks']:
-                            nb_graph.add_edge(block, block_name)
-                else:
-                    merge_code(nb_graph, block_name, c.source)
-
-    # Prepend all global code blocks before every other one
-    if global_block != "":
-        for block in nb_graph:
-            block_data = nb_graph.nodes(data=True)[block]
-            nx.set_node_attributes(nb_graph, {block: {
-                'source': global_block + '\n\n' + block_data['source']}})
+            # add node to DAG, adding tags and source code of notebook cell
+            if step_name not in nb_graph.nodes:
+                nb_graph.add_node(step_name, source=[c.source],
+                                  ins=set(), outs=set())
+                for _prev_step_name in tags['prev_steps']:
+                    nb_graph.add_edge(_prev_step_name, step_name)
+            else:
+                merge_code(nb_graph, step_name, c.source)
+
+        prev_step_name = step_name
```

Comment:
This must be inside the `else` statement. If it gets the `None` value then it won't merge the code. Two consecutive untagged blocks should be merged

**We definitely need a test for that!** Could you take a note for that?

```suggestion
            prev_step_name = step_name
```

Reply:
That was a great spot, and my bad for introducing this. It must have slipped while refactoring commits. Fixed :)